### PR TITLE
feat(EMS-2472): No PDF - Credit insurance cover - validation

### PR DESCRIPTION
**Pull request by Zainzzkk**
## Introduction :pencil2:
This PR adds validation for credit insurance cover page

## Resolution :heavy_check_mark:
- Added validation for credit insurance cover page
- Added tests
- Added post route
- Updated and added e2e tests

## Miscellaneous :heavy_plus_sign:
- Added maximum for credit insurance cover page textarea

### CODE REVIEW

File: e2e-tests/commands/insurance/your-buyer/complete-credit-insurance-cover-form.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * if hasHadCreditInsuranceCover, then selects yes radio and fills in conditional text area
+ * if not hasHadCreditInsuranceCover then selects no radio
+ * @param {Boolean} hasHadCreditInsuranceCover - should select yes or no radio
+ */
+const completeCreditInsuranceCoverForm = ({ hasHadCreditInsuranceCover = false }) => {
+  const fieldId = PREVIOUS_CREDIT_INSURANCE_COVER_WITH_BUYER;
+
+  if (hasHadCreditInsuranceCover) {
+    yesRadioInput().click();
+    cy.keyboardInput(field(fieldId).textarea(), 'test');
```
Comment (ttbarnes) [inline e2e-tests/commands/insurance/your-buyer/complete-credit-insurance-cover-form.js:20]: Can the value come from mock application, something more meaningful e.g "Mock cover with buyer"

File: src/ui/server/controllers/insurance/your-buyer/credit-insurance-cover/validation/rules/credit-insurance-cover-description.test.ts.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import { ERROR_MESSAGES } from '../../../../../../content-strings';
+import { RequestBody } from '../../../../../../../types';
+import generateValidationErrors from '../../../../../../helpers/validation';
+import inputValidation from '../../../../../../shared-validation/max-length';
+
+const {
+  INSURANCE: {
+    YOUR_BUYER: { PREVIOUS_CREDIT_INSURANCE_COVER_WITH_BUYER: FIELD_ID, HAS_PREVIOUS_CREDIT_INSURANCE_COVER_WITH_BUYER },
+  },
+} = FIELD_IDS;
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/credit-insurance-cover/validation/rules/credit-insurance-cover-description.test.ts.ts:12]: `INSURANCE_FIELD_IDS` 😛

File: src/ui/server/controllers/insurance/your-buyer/credit-insurance-cover/validation/rules/credit-insurance-cover-description.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import { FIELD_IDS } from '../../../../../../constants';
+import { ERROR_MESSAGES } from '../../../../../../content-strings';
+import { RequestBody } from '../../../../../../../types';
+import generateValidationErrors from '../../../../../../helpers/validation';
+import { objectHasProperty } from '../../../../../../helpers/object';
+import inputValidation from '../../../../../../shared-validation/max-length';
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/credit-insurance-cover/validation/rules/credit-insurance-cover-description.ts:6]: Can we rename to `maxLengthValidation` please, done this in another open PR

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/credit-insurance-cover/validation/credit-insurance-cover-validation.spec.js [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+  before(() => {
+    cy.completeSignInAndGoToApplication({}).then(({ referenceNumber: refNumber }) => {
+      referenceNumber = refNumber;
+
+      cy.startInsuranceYourBuyerSection({});
+
+      url = `${baseUrl}${ROOT}/${referenceNumber}${CREDIT_INSURANCE_COVER}`;
+
+      // TODO: EMS-2659 - use buyer commands to get here
```
[the remark this answers is not in the document]
Comment (Zainzzkk) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/credit-insurance-cover/validation/credit-insurance-cover-validation.spec.js:39]: it needs its own ticket and the rest of the flow to work too

File: e2e-tests/insurance/cypress/e2e/journeys/your-buyer/credit-insurance-cover/validation/credit-insurance-cover-validation.spec.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+      cy.submitAndAssertFieldErrors(
+        textareaField,
+        null,
+        0,
+        expectedErrorsCount,
+        ERRORS[fieldId].IS_EMPTY,
+      );
+    });
+
+    it(`should render a validation error and retain the submitted value when ${fieldId} is above the maximum`, () => {
```
[the remark this answers is not in the document]
Comment (Zainzzkk) [reply e2e-tests/insurance/cypress/e2e/journeys/your-buyer/credit-insurance-cover/validation/credit-insurance-cover-validation.spec.js:93]: removed this part

File: src/ui/server/controllers/insurance/your-buyer/credit-insurance-cover/validation/rules/credit-insurance-cover-description.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ */
+const creditInsuranceCoverDescriptionRule = (formBody: RequestBody, errors: object) => {
+  // if HAS_PREVIOUS_CREDIT_INSURANCE_COVER_WITH_BUYER radio is yes then check validation
+  if (formBody[HAS_PREVIOUS_CREDIT_INSURANCE_COVER_WITH_BUYER] === 'true') {
+    if (!objectHasProperty(formBody, FIELD_ID)) {
+      return generateValidationErrors(FIELD_ID, ERROR_MESSAGE.IS_EMPTY, errors);
+    }
+
+    return maxLengthValidation(formBody[FIELD_ID], FIELD_ID, ERROR_MESSAGE.ABOVE_MAXIMUM, errors, MAXIMUM);
+  }
```
Comment (ttbarnes) [inline src/ui/server/controllers/insurance/your-buyer/credit-insurance-cover/validation/rules/credit-insurance-cover-description.ts:35]: Do you think we should change this to:

```js

if (formBody[HAS_PREVIOUS_CREDIT_INSURANCE_COVER_WITH_BUYER] === 'true') {
  if (objectHasProperty(formBody, FIELD_ID)) {
    return maxLengthValidation(formBody[FIELD_ID], FIELD_ID, ERROR_MESSAGE.ABOVE_MAXIMUM, errors, MAXIMUM);
  }

  return emptyFieldValidation(FIELD_ID, ERROR_MESSAGE.IS_EMPTY, errors);
}
```

Have done this for shared "name" validation"

**Review by ttbarnes — APPROVED** [review]

**Pull request merged**